BTREE.CPP

vipinmankar opened the pull request:
//B tree cpp program
#include <iostream>
using namespace std;
class btnode{
    public :
    int *value;
    int t;
    btnode **child;
    int n;
    bool leaf;
    
    btnode(int t1,bool leaf1){
        t=t1;
        leaf=leaf1;
        value=new int[2*t-1];
        child=new btnode *[2*t];
        n=0;
    }
    void split(int i,btnode *y){
        btnode *z=new btnode(y->t,y->leaf);
        z->n=t-1;
        for (int j = 0; j < t-1; j++)
            z->value[j] = y->value[j+t];
        
        
        if (y->leaf == false)
        {
            for (int j = 0; j < t; j++)
                z->child[j] = y->child[j+t];
        }
        
        
        y->n = t - 1;
        
        
        for (int j = n; j >= i+1; j--)
            child[j+1] = child[j];
        
        child[i+1] = z;
        
        for (int j = n-1; j >= i; j--)
            value[j+1] = value[j];
        
        value[i] = y->value[t-1];
        
        n = n + 1;  }
    void insertnonfull(int k){
        int i = n-1;
        
        if (leaf == true)
        {
            while (i >= 0 && value[i] > k)
            {
                value[i+1] = value[i];
                i--;
            }
            
            value[i+1] = k;
            n = n+1;
        }
        else
        {
            while (i >= 0 && value[i] > k)
                i--;
            
            if (child[i+1]->n == 2*t-1)
            {
                split(i+1, child[i+1]);
                
                if (value[i+1] < k)
                    i++;
            }
            
        }
        
    }
    void traverse(){
        int i;
        for (i = 0; i < n; i++)
        {
            if (leaf == false)
                child[i]->traverse();
            cout <<"\n"<< value[i];
        }
        
        
        if (leaf == false)
            child[i]->traverse();
    }
    friend class btree;
};
class btree{
public:
    btnode  *root;
    int t,temp;
    btree(int t1){
        root=NULL;
        t=t1;
    }
    void traverse()
    {  if (root != NULL) root->traverse(); }
    void insert(){
        cout<<" "<<"Enter the number to add"<<"\n" ;
        cin>>temp;
        if(root==NULL){
            root=new btnode(t,true);
            root->value[0]=temp;
            root->n=1;
            
        }
        else{
            if(root->n==(2*t-1)){
                btnode *s=new btnode(t,false);
                s->child[0]=root;
                s->split(0,root);
                int i=0;
                if (s->value[0] < temp)
                    i++;
                s->child[i]->insertnonfull( temp);
                root = s;
                
            }
            
            else{
                root->insertnonfull( temp);
            }
        }
    }
    
};

int main()
{
    btree t(3);
    int opt;
    while(1){
        cout<<" \npress 1 for insert \npress 2 for traverse \npress 3 for Exit ";
        cin>>opt;
        switch(opt){
            case 1:
                t.insert();
                break;
            case 2:
                t.traverse();
                break;
            case 3:
                exit(0);
        }
    }
    return 0;
}

//Sample input :
10 21 31 41 51 
output 
10 21 31 41 51
 //Sample input :
70 31 31 21 51 
output 
21 21 31 51 70

raghav-dalmia commented on BTree.cpp:
```diff
@@ -0,0 +1,162 @@
+    return 0;
+}
+//Sample input :
+10 21 31 41 51 
```
Comment these input/ output

raghav-dalmia commented on BTree.cpp:
```diff
@@ -0,0 +1,162 @@
+        t=t1;
+    }
+    void traverse()
+    {  if (root != NULL) root->traverse(); }
```
use proper indentation

raghav-dalmia commented on BTree.cpp:
```diff
@@ -0,0 +1,162 @@
+                child[i]->traverse();
+            cout <<"\n"<< value[i];
+        }
+        
```
remove extra lines

raghav-dalmia commented on BTree.cpp:
```diff
@@ -0,0 +1,162 @@
+        n=0;
+    }
+    void split(int i,btnode *y){
+        btnode *z=new btnode(y->t,y->leaf);
```
use proper variable names

raghav-dalmia commented:
use the template for PR description, squash commits, add comments, give a small description about code at the beginning of code.

vipinmankar commented:
Thank you for viewing my work and for your valuable suggestions.
I will try to complete the  assigned work at the earliest.

On Tue, 3 Mar, 2020, 12:19 AM Raghav Dalmia, <notifications@github.com>
wrote:

> use the template for PR description, squash commits, add comments, give a
> small description about code at the beginning of code.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/jainaman224/Algo_Ds_Notes/pull/1894?email_source=notifications&email_token=AKJ3VJZKFZN4JHG7RUYYBDTRFP5VDA5CNFSM4K7SCG2KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOENQPRRI#issuecomment-593557701>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKJ3VJ4CEX7Y6YKMYWEHZG3RFP5VDANCNFSM4K7SCG2A>
> .
>

vipinmankar commented on BTree.cpp:
```diff
@@ -1,7 +1,7 @@
 
 
 // cpp program of bTree  ton insert and traverse the tree ,BTree is tree data structure which use to
- //indexing of database.
```
i

vipinmankar commented:
@raghav-dalmia i have made the changes as you mention please review it.
Thank you

vipinmankar commented:
@raghav-dalmia is there is any chances i have to make please let me know .

jainaman224 commented:
Fix issue template. It is compulsory.

jainaman224 commented on BTree.cpp:
```diff
@@ -0,0 +1,184 @@
+
+#include <iostream>
+using namespace std;
+class create_tree_node{
```
Add new line above this.

somya-kapoor replied:
Leave a line before it

jainaman224 commented on BTree.cpp:
```diff
@@ -0,0 +1,184 @@
+        child=new create_tree_node *[2*no_of_key];
+        count_of_keys=0;
+    }
+    //Function to split note and creayte new node
```
Add space after `//`.

somya-kapoor replied:
Add space after //.

jainaman224 commented on BTree.cpp:
```diff
@@ -0,0 +1,184 @@
+        for (int j = 0; j < no_of_key-1; j++)
+            z->Key[j] = y->Key[j+no_of_key];
+        
+         // Copy the last t children of y to z
```
Inconsistent indentation.

somya-kapoor replied:
Inconsistent indentation.

jainaman224 commented on BTree.cpp:
```diff
@@ -0,0 +1,184 @@
+
```
Add this file to a folder.

vipinmankar commented:
@raghav-dalmia sir @jainaman224 sir i have made the required changes please merge this PR.

somya-kapoor commented on BTree.cpp:
```diff
@@ -0,0 +1,184 @@
+    int count_of_keys;
+    bool leaf;
+    
+    create_tree_node(int t1,bool leaf1){
```
Leave a space after ,

somya-kapoor commented on BTree.cpp:
```diff
@@ -0,0 +1,184 @@
+    bool leaf;
+    
+    create_tree_node(int t1,bool leaf1){
+        no_of_key=t1;
```
Add space around the operator

somya-kapoor commented on BTree.cpp:
```diff
@@ -0,0 +1,184 @@
+    
+    create_tree_node(int t1,bool leaf1){
+        no_of_key=t1;
+        leaf=leaf1;
```
Add space around the operator. Do this in whole file

somya-kapoor commented on BTree.cpp:
```diff
@@ -0,0 +1,184 @@
+
```
There are 2 files with the same code, delete one of them.

vipinmankar replied:
@somya-kapoor yes maam , by tomorrow morning i will commit it with all required  changes.

somya-kapoor commented:
@vipinmankar Are you working on it?

somya-kapoor commented:
@vipinmankar There is no file in the PR

vipinmankar commented:
@somya-kapoor  i have added the file as it was not there  ,as you mention earlier that my repo contain 2 file of same name so after that  it i had deleted those from master branch and create a new branch made the new  pr but it  was not showing pr .So now i have again added it to the master and it can be now be viewed in pr. Hence I request you to review it . 
Thank you

somya-kapoor commented:
@vipinmankar You have created 2 PR for the same issue #1894 and #2686. Please close on of them.

somya-kapoor commented:
@vipinmankar Are you working on it?
